updated recording page

Mohamed-Dhouioui opened the pull request:
I added an explanation of how to record and playback using a Python client. The recording format may need a more detailed description too.

Ragzouken commented on source/concepts/recording.rst:
```diff
@@ -69,8 +68,42 @@ In order to send both streams together, provide the two file paths separated by
+    # Define the .traj and .state file names and paths
+    traj_path = 'simulation_recording.traj'
+    state_path = 'simulation_recording.state'
+    # Pass the runner to the recording function
```
This comment is inaccurate, maybe a typo? Might be better to say the it "creates a client that connects to a server and records until the server stops" or something

hjstroud commented on source/concepts/recording.rst:
```diff
@@ -69,8 +68,42 @@ In order to send both streams together, provide the two file paths separated by
+    traj_path = 'simulation_recording.traj'
+    state_path = 'simulation_recording.state'
+    # create a recording from a server and save it to the files
+    record_from_server("localhost:38801", traj_path, state_path)
```
May be worth adding a line after the code block to point people towards the tutorials that demonstrate how to record using NanoVer, i.e. the `recording_and_replaying.ipynb` notebook of the basics section of the tutorials (and linking to it?)

hjstroud commented on source/concepts/recording.rst:
```diff
@@ -69,8 +68,42 @@ In order to send both streams together, provide the two file paths separated by
+    recording_runner.next()
+    # Close the runner
+    recording_runner.close()
+
```
Again, may be worth adding a line directing people towards the `recording_and_replaying.ipynb` notebook of the basics section of the tutorials, which demonstrates how to do this. Something along the lines of "See our `recording_and_replaying.ipynb`(url) for an example of how to replay recorded simulations."

hjstroud reviewed (commented):
I think this adds the required information quite succinctly. It may be worth rearranging the page so that the python recording and replaying sections appear _before_ the Rust recording and replaying sections (as the Python server is the primary server that we support), and moving some of the general information in the Rust section to the Python section instead (if it applies).

rhoslynroebuck commented:
I agree with Harry's comment, I think the Python section should come before the Rust section. Then the page needs reading through from start to finish to make sure there is a nice flow. As Harry said, the Rust section will need rewording because it was written to be standalone, rather than after a section about recording in Python. 

Another general comment is that I think there's a lot of technical jargon that wouldn't be clear to a new reader. I think these terms either need to be explained on-the-fly on this page, or linked to other places in the documentation where they are explained. Also, I think some terms need a bit more clarity. For example, a "recording" can refer to different things. For example, the line "The [nanover.mdanalysis](https://github.com/IRL2/nanover-docs/python/nanover.mdanalysis.html#module-nanover.mdanalysis) module allows to read a trajectory recording as an [MDAnalysis Universe](https://userguide.mdanalysis.org/stable/universe.html#universe)". What type of trajectory recording file are you referring to here? Because the trajectory file format that we record in NanoVer is our own, not a typical one that a computational chemistry program uses by default.

Mohamed-Dhouioui commented:
Thank you for the feedback, I have made these changes based on your comments:

- Rearranged the page to be more consistent with NanoVer current priorities.
- Added links and references to relevant tutorials/docs for further info.
- Edited the rust recording section and moved some information to the python section

Please have a read-through to check if all information is correctly described and if anything is missing feel free to report it.

Mohamed-Dhouioui commented:
Thank you for the feedback, I have made these changes based on your comments:

Rearranged the page to be more consistent with NanoVer current priorities.
Added links and references to relevant tutorials/docs for further info.
Edited the rust recording section and moved some information to the python section
Please have a read-through to check if all information is correctly described and if anything is missing feel free to report it.

hjstroud commented on source/concepts/recording.rst:
```diff
@@ -25,6 +25,81 @@ Each stream is stored in a separate file.
+Recording format
+----------------
+
+The current version of the file format is version 2.
```
I think we can remove this line, as the current version is already stated below in the bullet points

Mohamed-Dhouioui replied:
Done in commit [52dc633](https://github.com/IRL2/nanover-docs/pull/113/commits/52dc633153e8d43308b50314cf888181b9394a23)

hjstroud commented on source/concepts/recording.rst:
```diff
@@ -25,6 +25,81 @@ Each stream is stored in a separate file.
+This message contains two fields: the frame index and the frame itself.
+The frame index is generally an integer that gets incremented each time the server register a frame to broadcast.
+However, its value is only significant when it is 0 as it means the frame needs to be reset;
+for instance because the server loaded a new simulation. The frame is a :ref:`FrameData <traj-and-frames>`.
```
```suggestion
for instance because the server loaded a new simulation. The frame itself is defined by the :`FrameData <traj-and-frames>` class.
```

Mohamed-Dhouioui replied:
Done in commit [52dc633](https://github.com/IRL2/nanover-docs/pull/113/commits/52dc633153e8d43308b50314cf888181b9394a23).

hjstroud commented on source/concepts/recording.rst:
```diff
@@ -25,6 +25,81 @@ Each stream is stored in a separate file.
+
+    from nanover.omni import OmniRunner
+    from nanover.omni.playback import PlaybackSimulation
+    simulation_recording = PlaybackSimulation(name='simulation-recording', traj='files/simulation_recording.traj',
```
It may be worth splitting the name and traj parts over two lines so that the example doesn't run over the edge of the page

Mohamed-Dhouioui replied:
Done in commit [52dc633](https://github.com/IRL2/nanover-docs/pull/113/commits/52dc633153e8d43308b50314cf888181b9394a23).

hjstroud commented on source/concepts/recording.rst:
```diff
@@ -25,6 +25,81 @@ Each stream is stored in a separate file.
+    simulation_recording = PlaybackSimulation(name='simulation-recording', traj='files/simulation_recording.traj',
+                                           state='files/simulation_recording.state')
+    # Create a runner for the simulation
+    recording_runner = OmniRunner.with_basic_server(simulation_recording, name='simulation-recording-server')
```
Again, split over multiple lines so the code example doesn't run off of the page

Mohamed-Dhouioui replied:
Done in commit [52dc633](https://github.com/IRL2/nanover-docs/pull/113/commits/52dc633153e8d43308b50314cf888181b9394a23).

hjstroud commented on source/concepts/recording.rst:
```diff
@@ -70,7 +139,7 @@ In order to send both streams together, provide the two file paths separated by
 
 
-Reading recordings using python
+Reading recordings using mdanalysis in python
```
```suggestion
Reading NanoVer recordings using MDAnalysis in python
```
I think it would be good to make it clear throughout this page that we are talking about NanoVer recordings, i.e. recordings made using the inbuilt recording functionality of NanoVer (as it is possible to record trajectories in other ways)

Mohamed-Dhouioui replied:
Done in commit [52dc633](https://github.com/IRL2/nanover-docs/pull/113/commits/52dc633153e8d43308b50314cf888181b9394a23).

hjstroud reviewed (changes requested):
This is looking better, things seem to be in the right order now and there are good links to tutorial notebooks and other pages that supplement what is talked about on this page.

I agree with the comment that @rhoslynroebuck made: it is really important on this page to be crystal clear about what type of recording procedure we are talking about. This page is about explaining how to record simulations using the inbuilt functionality provided by NanoVer, so maybe it should be called "Recording simulations using NanoVer" or something similar, and there should be a sentence in the first paragraph that states that this page details how to record NanoVer sessions using the inbuilt functionality of NanoVer.

Perhaps it would be more clear to refer to trajectory recording files as NanoVer trajectory recording files, and state recording files as NanoVer state recording files, to make it unmistakeable what is being referenced at each point.

In general I think the page needs to be read through and clarified so that it flows better, for example the sentence "The `nanover.mdanalysis` module allows to read a trajectory as an MDAnalysis universe" doesn't make sense and needs to be reworded.

I have left some individual comments throughout the code too, it would be good to address these. @Mohamed-Dhouioui you have already addressed some of these, so I will close the ones already addressed as resolved.

Mohamed-Dhouioui commented:
I have renamed the page title as suggested by Harry, added a sentence in the first paragraph stating the recording purpose, changed the referring of trajectory and shared state into NanoVer recording and NanoVer shared state, and added a phrase explaining what an MDanalysis Universe is.

hjstroud reviewed (approved):
I have reformatted the initial paragraphs to condense them slightly and improve clarity, and made a few minor changes to headings, wording and code examples. I think this is now ready to be pushed to main!
